### PR TITLE
[POC] [Workflow] Made StateMachine supports only single state marking

### PR DESCRIPTION
**Pull request by HeahDude**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/pull/19629
| License       | MIT
| Doc PR        | TODO

Since StateMachine cannot be in one place at the time it does not make sense to allow multiple marking. Actually we default to the right SingleStateMarkingStore, but we can't prevent it with any interface.

This PR makes `Marking` and `MarkingStore` abstract so any implementation can use the component mechanisms by defining a marking strategy using `Marking` constants in the `MarkingStore` constructor.

Hence we need two marking classes `SingleStateMarking` and `MultipleStateMarking` so every internal piece of the workflow knows how to behave regarding marking.

Also we only need one `PropertyAccessMarkingStore` which use the same decency for its logic but can handle both strategies.

Most of the abstract class are public and can easily be overridden.

Wdyt?

### CODE REVIEW

**Comment by stof** [comment]
This prevents using any other custom implementation of the MarkingStore return astate with a single place, so -1 for that.

**Comment by xabbuh** [comment]
Maybe we could think about introducing dedicated marker interfaces for marking stores that support or do not support multiple states at the same time.

**Comment by HeahDude** [comment]
Ok I think I've got an idea.

Status: needs work

**Comment by HeahDude** [comment]
Ok I've updated the PR and the description, I'm finishing some tests, but this is ready to review as POC.

Maybe it will inspire someone for something better :)

File: src/Symfony/Component/Workflow/MarkingStore/AbstractMarkingStore.php [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+     */
+    abstract protected function getSingleStateMarking($subject);
+
+    /**
+     * Sets a SingleStateMarking to a subject.
+     *
+     * @param object             $subject A subject
+     * @param SingleStateMarking $marking
+     *
+     * @return SingleStateMarking The marking
```
Comment (stof) [inline src/Symfony/Component/Workflow/MarkingStore/AbstractMarkingStore.php]: setters have no reason to return the marking (the return value is ignored anyway)

File: src/Symfony/Component/Workflow/MultipleStateMarking.php [hunk 2]
```diff
@@ -30,22 +33,48 @@ public function __construct(array $representation = array())
...
     public function mark($place)
     {
-        $this->places[$place] = 1;
+        if (!isset($this->places[$place])) {
+            $this->places[$place] = 1;
+
+            return;
+        }
+
+        $this->places[$place] = $this->places[$place] + 1;
```
Comment (stof) [inline src/Symfony/Component/Workflow/MultipleStateMarking.php]: ``+= 1`` or ``++``

File: src/Symfony/Component/Workflow/Registry.php [hunk 3]
```diff
@@ -30,6 +30,14 @@ public function add(Workflow $workflow, $className)
         $this->workflows[] = array($workflow, $className);
     }
 
+    /**
+     * @param      $subject
+     * @param null $workflowName
```
Comment (stof) [inline src/Symfony/Component/Workflow/Registry.php]: wrong phpdoc

File: src/Symfony/Component/Workflow/SingleStateMarking.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    public function __construct($place = null)
+    {
+        if (null !== $place && !is_scalar($place)) {
+            throw new InvalidArgumentException(sprintf('"%s" instances only accept scalar or null as single place, but got "%s".', __CLASS__, is_object($place) ? get_class($place) : gettype($place)));
+        }
+
+        if ($place) {
+            $this->mark($place);
+        } else {
+            $this->place = array();
```
Comment (stof) [inline src/Symfony/Component/Workflow/SingleStateMarking.php]: useless as the property is already initialized

**Comment by stof** [comment]
@lyrixx could you have a look at this one ?

**Comment by HeahDude** [comment]
@stof thank you for your comments, they're now addressed, I might have fixed another bug in 2289c55.

**Comment by HeahDude** [comment]
Reviewing my code on GH, I think it might be worth having `Marking` and `MarkingStore` abstract and get rid of the interfaces.

**Comment by HeahDude** [comment]
I've updated this PR so it has only one clean commit and a better description.

Status: needs review

**Comment by HeahDude** [comment]
Failures unrelated.

File: src/Symfony/Component/Workflow/SingleStateMarking.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+/**
+ * SingleStateMarking contains the place of one token.
+ *
+ * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+class SingleStateMarking extends Marking
+{
+    /**
+     * @param mixed $place A scalar as only place or null
```
Comment (ro0NL) [inline src/Symfony/Component/Workflow/SingleStateMarking.php]: Should be `string|null`

File: src/Symfony/Component/Workflow/SingleStateMarking.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    public function __construct($place = null)
+    {
+        if (null !== $place && !is_scalar($place)) {
+            throw new InvalidArgumentException(sprintf('"%s" instances only accept scalar or null as single place, but got "%s".', __CLASS__, is_object($place) ? get_class($place) : gettype($place)));
```
Comment (ro0NL) [inline src/Symfony/Component/Workflow/SingleStateMarking.php]: Not sure this safeguard is actually needed.

File: src/Symfony/Component/Workflow/SingleStateMarking.php [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        }
+
+        if ($place) {
```
Comment (ro0NL) [inline src/Symfony/Component/Workflow/SingleStateMarking.php]: Should be `if(null !== $place) {`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 8]
```diff
@@ -431,13 +431,16 @@ private function registerWorkflowConfiguration(array $workflows, ContainerBuilde
...
+                $markingStoreDefinition = new Reference($workflow['marking_store']['service']);
+            } else {
+                $markingStoreDefinition = new DefinitionDecorator('workflow.marking_store.property_access');
                 foreach ($workflow['marking_store']['arguments'] as $argument) {
                     $markingStoreDefinition->addArgument($argument);
                 }
-            } elseif (isset($workflow['marking_store']['service'])) {
-                $markingStoreDefinition = new Reference($workflow['marking_store']['service']);
+
+                $markingStrategy = 'state_machine' === $type ? 'single_state' : $workflow['marking_store']['type'];
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: `$markingStrategy = 'state_machine' === $type ? Marking::STRATEGY_SINGLE_STATE : Marking::STRATEGY_MULTIPLE_STATE;`?

File: src/Symfony/Component/Workflow/Exception/InvalidMarkingException.php [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+ */
+class InvalidMarkingException extends LogicException
+{
+    /**
+     * @param string $expectedClass
+     * @param mixed  $marking
+     */
+    public function __construct($expectedClass, $marking)
+    {
+        $this->message = sprintf('Marking must be an instance of "%", but got "%"', $expectedClass, is_object($marking) ? get_class($marking) : gettype($marking));
```
Comment (ro0NL) [inline src/Symfony/Component/Workflow/Exception/InvalidMarkingException.php]: `parent::__construct(...)`

File: src/Symfony/Component/Workflow/Marking.php [hunk 10]
```diff
@@ -12,41 +12,45 @@
...
+ * by places with tokens.
  *
  * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ * @author Jules Pietri <jules@heahprod.com>
  */
-class Marking
+abstract class Marking
 {
-    private $places = array();
+    const STRATEGY_SINGLE_STATE = 'single_state';
```
Comment (ro0NL) [inline src/Symfony/Component/Workflow/Marking.php]: These constants belong to the `MarkingStore` class imo.

File: src/Symfony/Component/Workflow/Marking.php [hunk 11]
```diff
@@ -12,41 +12,45 @@
...
- * Marking contains the place of every tokens.
+ * A base Marking which contains the state of the
+ * state of a Workflow or a StateMachine a representation
+ * by places with tokens.
  *
  * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ * @author Jules Pietri <jules@heahprod.com>
  */
-class Marking
+abstract class Marking
```
Comment (ro0NL) [inline src/Symfony/Component/Workflow/Marking.php]: Not sure this class is really needed.. why not a pointer interface and 2 final implementations instead?

**Comment by lyrixx** [comment]
@ro0NL Hi. Thanks for your review, But this is a POC. So please **don't** comment about CS / PHPDocs etc... @HeahDude will fix all those tiny details later if the PR is accepted.

**Comment by lyrixx** [comment]
Hello @HeahDude 

I understand why you made this PR and I agree a pitfall may exist for newcomers.

But I don't like your current implementation:

1/ There are too many code added, it adds too many complexity
2/ You renamed place into state, this is wrong

It could be solved in the full stack framework in a simpler way: just check the framework type with the type marking store type.
It will not solve all errors since it's possible to use a custom service for the marking store. But I think if people implement a custom marking store, they know what they are doing.

I don't implemented what I said, because it's not perfect and because the default configuration is error proof.

If you can find a much simpler solution to implement this kind of validation in the composant, I will be glad to merge it. but for now I'm -1

**Comment by HeahDude** [comment]
@lyrixx I understand. Thank you for your feedback, I'll try to look at it again with this in mind :)

**Comment by HeahDude** [comment]
I'll come with something else :)

Status: needs work

**Comment by lyrixx** [comment]
I'm sorry, but I have to close this PR as the current implementation is too complex
(I have talked with @HeahDude IRL about this PR)

**Pull request closed without merging**